### PR TITLE
Add support for selective finetune (freeze parameters by regex from config file)

### PR DESCRIPTION
**Pull request by HarshTrivedi**
- Add support in fine_tune to selectively tune
- Add tests for selective fine tuning.
- Allow for turning off gradients in train command.

This PR adds support in `fine-tune` command to freeze parameters layers by passing regex in config file. In config file passed in fine-tune command:
```
"trainer": {
    ...
    "no_grad": ["*conv*", ".*text_embedder*"]
}
```
The above will freeze `conv` layers and `text_embedder`s but not others parameters.
I believed best place to allow `no_grad` key setting accessible to is within `trainer` key and `trainer` is also used in `train` command. So I allowed turning of gradients via `no_grad` in train command as well.  

Although I primarily intended this selective turning off of gradients for fine-tune command, it can be useful in `train` command as well : - eg. If one's modules parameters are loaded / transferred from some other pretrained model and want to freeze it there.

More context: Issue (#1298 )

### CODE REVIEW

File: allennlp/commands/fine_tune.py [hunk 1]
```diff
@@ -166,6 +166,12 @@ def fine_tune_model(model: Model,
     test_data = all_datasets.get('test')
 
     trainer_params = params.pop("trainer")
+    nograd_regex_list = trainer_params.pop("no_grad", ())
+    if nograd_regex_list:
+        nograd_regex = "(" + ")|(".join(nograd_regex_list) + ")"
```
Comment (joelgrus) [inline allennlp/commands/fine_tune.py]: this feels error-prone to me, I'd rather just 

```
no_grad_regexes = trainer_params.pop("no_grad", ())
for name, parameter in model.named_parameters():
    if any(re.search(regex, name) for regex in no_grad_regexes):
        parameter.requires_grad_(False)
```

it's cleaner and I can't imagine the time difference being noticeable

File: allennlp/commands/train.py [hunk 2]
```diff
@@ -282,6 +282,12 @@ def train_model(params: Params,
     test_data = all_datasets.get('test')
 
     trainer_params = params.pop("trainer")
+    nograd_regex_list = trainer_params.pop("no_grad", ())
+    if nograd_regex_list:
+        nograd_regex = "(" + ")|(".join(nograd_regex_list) + ")"
+        for name, parameter in model.named_parameters():
+            if re.search(nograd_regex, name):
+                parameter.requires_grad_(False)
```
Comment (joelgrus) [inline allennlp/commands/train.py]: same comment here

File: allennlp/commands/fine_tune.py [hunk 3]
```diff
@@ -10,7 +10,7 @@
 import logging
 import os
 from copy import deepcopy
-
+import re
```
Comment (joelgrus) [inline allennlp/commands/fine_tune.py]: nit: I like to leave blank lines between the grouped imports: (standard library) -> (external libraries) -> (allennlp modules)

File: allennlp/commands/train.py [hunk 4]
```diff
@@ -37,7 +37,7 @@
 import logging
 import os
 from copy import deepcopy
-
+import re
```
Comment (joelgrus) [inline allennlp/commands/train.py]: same nit

File: allennlp/tests/commands/fine_tune_test.py [hunk 5]
```diff
@@ -1,8 +1,13 @@
 # pylint: disable=invalid-name,no-self-use
 import argparse
-
+import re
+import shutil
+import pytest
```
Comment (joelgrus) [inline allennlp/tests/commands/fine_tune_test.py]: same nit: blank line before and after `pytest`

File: allennlp/tests/commands/train_test.py [hunk 6]
```diff
@@ -2,7 +2,8 @@
 import argparse
 from typing import Iterable
 import os
-
+import shutil
+import re
```
Comment (joelgrus) [inline allennlp/tests/commands/train_test.py]: same nit: blank line before `import pytest`

File: allennlp/tests/commands/fine_tune_test.py [hunk 7]
```diff
@@ -50,3 +55,36 @@ def test_fine_tune_fails_without_required_args(self):
...
+            params["trainer"]["no_grad"] = regex_list
+            shutil.rmtree(self.serialization_dir, ignore_errors=True)
+            tuned_model = fine_tune_model(model=original_model,
+                                          params=params,
+                                          serialization_dir=self.serialization_dir)
+            # If regex is matched, parameter name should have requires_grad False
+            # If regex is matched, parameter name should have same requires_grad
+            # as the originally loaded model
+            if regex_list:
+                nograd_regex = "(" + ")|(".join(regex_list) + ")"
```
Comment (joelgrus) [inline allennlp/tests/commands/fine_tune_test.py]: I'm more ok with this being in the test, but I'd still prefer the other way

File: allennlp/tests/commands/train_test.py [hunk 8]
```diff
@@ -232,3 +233,54 @@ def test_train_with_test_set(self):
...
+                       [".*text_field_embedder.*", ".*encoder.*"]]
+        for regex_list in regex_lists:
+            params = params_get()
+            params["trainer"]["no_grad"] = regex_list
+            shutil.rmtree(serialization_dir, ignore_errors=True)
+            model = train_model(params, serialization_dir=serialization_dir)
+            # If regex is matched, parameter name should have requires_grad False
+            # Or else True
+            if regex_list:
+                nograd_regex = "(" + ")|(".join(regex_list) + ")"
```
Comment (joelgrus) [inline allennlp/tests/commands/train_test.py]: same

File: allennlp/commands/fine_tune.py (hunk 1 above)
Comment (HarshTrivedi) [reply allennlp/commands/fine_tune.py]: Sure, will change that.

File: allennlp/commands/train.py (hunk 2 above)
Comment (HarshTrivedi) [reply allennlp/commands/train.py]: sure.

File: allennlp/commands/fine_tune.py (hunk 3 above)
Comment (HarshTrivedi) [reply allennlp/commands/fine_tune.py]: Ohh!! I didn't realise the purpose of blanks...

File: allennlp/commands/train.py (hunk 4 above)
Comment (HarshTrivedi) [reply allennlp/commands/train.py]: sure.

File: allennlp/tests/commands/fine_tune_test.py (hunk 5 above)
Comment (HarshTrivedi) [reply allennlp/tests/commands/fine_tune_test.py]: ok

File: allennlp/tests/commands/train_test.py (hunk 6 above)
Comment (HarshTrivedi) [reply allennlp/tests/commands/train_test.py]: ok

File: allennlp/tests/commands/fine_tune_test.py (hunk 7 above)
Comment (HarshTrivedi) [reply allennlp/tests/commands/fine_tune_test.py]: no problem, i can change that.

**Comment by HarshTrivedi** [comment]
True! Sure, I can add that in logs. thanks

**Comment by DeNeutoy** [comment]
Also FYI - thanks for your contributions to allennlp - your PRs and conduct are exemplary and i'm glad that you find the library useful.

**Comment by HarshTrivedi** [comment]
@DeNeutoy Thank you! allennlp is being very helpful to me in what I am currently doing and am sure others who try will find the same. I am very glad to contribute in allennlp as well : ) 👍

**Comment by HarshTrivedi** [comment]
Btw, do we want to log both frozen and non-frozen parameters? Most of the times most of the parameters will be non-frozen. And with all those parameters logged it looks quite messy on the terminal. How about logging only Frozen ones?

This is from one test: link. 
```
18:17:25 - INFO - allennlp.commands.train - Parameters without gradient (Frozen) : ['text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._char_embedding_weights', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_0.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_0.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_1.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_1.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_2.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_2.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_3.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_3.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_4.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder.char_conv_4.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._highways._layers.0.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._highways._layers.0.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._highways._layers.1.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._highways._layers.1.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._projection.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._token_embedder._projection.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_0.input_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_0.state_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_0.state_linearity.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_0.state_projection.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_0.input_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_0.state_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_0.state_linearity.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_0.state_projection.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_1.input_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_1.state_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_1.state_linearity.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.forward_layer_1.state_projection.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_1.input_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_1.state_linearity.weight', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_1.state_linearity.bias', 'text_field_embedder.token_embedder_elmo._elmo._elmo_lstm._elmo_lstm.backward_layer_1.state_projection.weight', 'text_field_embedder.token_embedder_elmo._elmo.scalar_mix_0.gamma', 'text_field_embedder.token_embedder_elmo._elmo.scalar_mix_0.scalar_parameters.0', 'text_field_embedder.token_embedder_elmo._elmo.scalar_mix_0.scalar_parameters.1', 'text_field_embedder.token_embedder_elmo._elmo.scalar_mix_0.scalar_parameters.2', 'text_field_embedder.token_embedder_tokens.weight', 'encoder._module.weight_ih_l0', 'encoder._module.weight_hh_l0', 'encoder._module.bias_ih_l0', 'encoder._module.bias_hh_l0', 'encoder._module.weight_ih_l0_reverse', 'encoder._module.weight_hh_l0_reverse', 'encoder._module.bias_ih_l0_reverse', 'encoder._module.bias_hh_l0_reverse', 'encoder._module.weight_ih_l1', 'encoder._module.weight_hh_l1', 'encoder._module.bias_ih_l1', 'encoder._module.bias_hh_l1', 'encoder._module.weight_ih_l1_reverse', 'encoder._module.weight_hh_l1_reverse', 'encoder._module.bias_ih_l1_reverse', 'encoder._module.bias_hh_l1_reverse', 'tag_projection_layer._module.weight', 'tag_projection_layer._module.bias', 'crf.transitions', 'crf._constraint_mask', 'crf.start_transitions', 'crf.end_transitions']
18:17:25 - INFO - allennlp.commands.train - Parameters with gradient    (Tunable): []
```

**Comment by DeNeutoy** [comment]
Put the individual parameters in separate logging statements in a for loop.

**Comment by HarshTrivedi** [comment]
This should be good now.

File: allennlp/commands/fine_tune.py [hunk 9]
```diff
@@ -180,6 +181,24 @@ def fine_tune_model(model: Model,
...
+    no_grad_regexes = trainer_params.pop("no_grad", ())
+
+    nograd_parameter_names = []
+    grad_parameter_names = []
+    for name, parameter in model.named_parameters():
+        if any(re.search(regex, name) for regex in no_grad_regexes):
+            parameter.requires_grad_(False)
+            nograd_parameter_names.append(name)
+        else:
+            grad_parameter_names.append(name)
```
Comment (rulai-huajunzeng) [inline allennlp/commands/fine_tune.py]: There is only one problem here, if the parameter's requires_grad is already False, for example a non-trainable embedding, the log will show that it is tunable. Not a very big problem but it looks a bit confusing sometimes.

Comment (HarshTrivedi) [reply allennlp/commands/fine_tune.py]: Thank You for catching that! I agree it would be confusing. Will fix that.
